web/app/view/CustomNumberField.js

```diff
@@ -20,30 +20,79 @@ Ext.define('Traccar.view.CustomNumberField', {
     xtype: 'customNumberField',
 
     beforeEl: '<div style="width:100%;display:inline-table;">',
-    unitEl: '<div style="display:table-cell;padding-left:10px;vertical-align:middle;">',
+    unitEl: '<div id="numberUnitEl" style="display:table-cell;padding-left:10px;vertical-align:middle;width:1px;white-space:nowrap;">',
 
     constructor: function (config) {
-        var unit;
-        if (config.dataType === 'speed') {
-            unit = Traccar.app.getPreference('speedUnit', 'kn');
-            config.beforeSubTpl = this.beforeEl;
-            config.afterSubTpl = this.unitEl + Ext.getStore('SpeedUnits').findRecord('key', unit).get('name') + '</div></div>';
+        var unitName = '';
+        if (config.dataType) {
+            config.beforeBodyEl = this.beforeEl;
+            switch (config.dataType) {
+                case 'speed':
+                    config.units = {};
+                    config.units.getStore = function () {
+                        return Ext.getStore('SpeedUnits');
+                    };
+                    config.units.getValue = function () {
+                        return Traccar.app.getPreference('speedUnit', 'kn');
+                    };
+                    unitName = Ext.getStore('SpeedUnits').findRecord('key', config.units.getValue()).get('name');
+                    break;
+                case 'distance':
+                    config.units = {};
+                    config.units.getStore = function () {
+                        return Ext.getStore('DistanceUnits');
+                    };
+                    config.units.getValue = function () {
+                        return Traccar.app.getPreference('distanceUnit', 'km');
+                    };
+                    unitName = Ext.getStore('DistanceUnits').findRecord('key', config.units.getValue()).get('name');
+                    break;
+                case 'frequency':
+                    if (!config.listeners) {
+                        config.listeners = {};
+                    }
+                    config.listeners.afterrender = function () {
+                        if (!this.units) {
+                            this.units = Ext.create({
+                                xtype: 'combobox',
+                                renderTo: 'numberUnitEl',
+                                store: 'TimeUnits',
+                                displayField: 'name',
+                                valueField: 'key',
+                                editable: false,
+                                numberField: this,
+                                value: 's',
+                                width: '70px',
```

Comment:
Unfortunately I did not managed avoid setting  width. It is needed to automatically fit width of cell.

![image](https://user-images.githubusercontent.com/5688080/30046935-b14e8ece-9227-11e7-8f1f-deb8a65fb9ee.png)